add rowRestriction for bigDiffy bq inputs

jrmcglynn opened the pull request:
This PR adds a `rowRestriction` argument to `bigDiffy` for BigQuery inputs. This allows for comparing partitions of natively partitioned BigQuery tables, but could be used in other ways as well. Moving BigQuery reads to `bigQueryStorage` was a necessity to implement this efficiently, but should be (faster and cheaper)[https://spotify.github.io/scio/io/BigQuery.html#bigquerytype-fromstorage] in addition.

`rowRestriction` was not implemented for avro inputs so a check was made to ensure that users do not pass in that combination of arguments.

jrmcglynn commented on ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala:
```diff
@@ -700,6 +707,10 @@ object BigDiffy extends Command with Serializable {
     val result = inputMode match {
       case "avro" =>
+        if(rowRestriction.isDefined) {
+          throw new NotImplementedError(s"rowRestriction is not implemented for avro inputs")
```
good call! lmk what you think of the updated exception

jrmcglynn commented on ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/BigDiffy.scala:
```diff
@@ -342,17 +342,21 @@ object BigDiffy extends Command with Serializable {
     ignoreNan: Boolean = false
-  ): BigDiffy[TableRow] =
+  ): BigDiffy[TableRow] = {
+    val restrictionCleaned = rowRestriction.map(_.replace('"', ' '))
```
are you thinking something like this? handles ", ', or `; and makes sure that the character being replaced is not in the middle of the line
```
val restrictionCleaned = rowRestriction.map(
      _.replaceAll("(^[\"'`])|([\"'`]$)", s"")
    )
```

jrmcglynn replied:
see my [update here](https://github.com/spotify/ratatool/pull/489/files#diff-8f727d03740a75cd83540588c8a261da38818e1b6d5aa13a9be5b85e53897a9eR340-R352) -- checks if the first and last characters are quote chars and are identical to each other, and removes if so

jrmcglynn commented on ratatool-diffy/src/test/scala/com/spotify/ratatool/diffy/BigDiffyTest.scala:
```diff
@@ -317,4 +317,23 @@ class BigDiffyTest extends PipelineSpec {
 
     exc.getMessage shouldBe "Output mode is GCS, but output abc is not a valid GCS location"
   }
+
```
I'm on vacation for the next week, but I can give this a shot when I get back

jrmcglynn commented:
I've successfully tested the following scenarios by triggering from the CLI locally:

- `rowRestriction` arg with wrapped quotes, `--rowRestriction="date=DATE '2021-10-01'"`
- `rowRestriction` arg with no quotes, `--rowRestriction=date=DATE'2021-10-01'`
- `rowRestriction` arg with a compound filter, `--rowRestriction="licensor='Universal Music Group' AND artist_name=‘Avicii'"`
- No `rowRestriction`

For each `rowRestriction` scenario I've confirmed that the filter was successfully applied on the input. The only time the filter was not successfully applied was when I did `--rowRestriction="date='2021-10-01'"`; apparently BigQuery's storage API is picky about needing the date string to be explicitly cast into a date type.

Let me know any further thoughts here!

idreeskhan reviewed (approved):
Looks good, can I merge?

idreeskhan commented:
Oh wait sorry one more thing, can you document the potential inputs for rowRestriction somewhere? Those are like 3 different types that you tested.

jrmcglynn commented:
@idreeskhan I added a section in the BigDiffy README, is that along the lines of what you were thinking?

I also modified the CLI docs to make it clear that it's an optional param and match Google's [doc's for this parameter](https://cloud.google.com/bigquery/docs/reference/storage/rpc/google.cloud.bigquery.storage.v1#google.cloud.bigquery.storage.v1.ReadSession.TableModifiers)
